pkg/virt-handler/notify-server/server.go

```diff
@@ -154,7 +154,7 @@ func RunServer(virtShareDir string, stopChan chan struct{}, c chan watch.Event,
 	case <-done:
 		log.Log.Info("notify server done")
 	case <-stopChan:
-		grpcServer.Stop()
+		grpcServer.GracefulStop()
```

Comment:
I think this is the right approach.

One thing that has me slightly concerned is that GracefulStop is an unbounded blocking action though. 

Maybe I'm just being over cautious, but what would you think about something like this? Basically, I just want us to know if if graceful shutdown ever blocks indefinitely rather than discovering in another odd way. 

```
gracefulShutdownTimeout := 30

isStopped := make(chan, struct{})
go func () {
     grpcServer.GracefulStop()
     close(isStopped)
}

t := time.NewTimer(gracefulShutdownTimeout * time.Second)
select {
case <- t.C:
    // if we can't stop the server gracefully after 'x' seconds, panic
    panic(fmt.Errorf("timed out attempting to stop grpc notify server after %d seconds", gracefulShutdownTimeout))
case <- stopped:
    t.Stop()
}
```

Reply:
Good idea, thanks David, on it, 
For example if reflector would have deadlock for some kind of bug,
eventChan would block after 100 events (this is the depth right of the chan right now) and this would detect it indeed.

Reply:
Updated PR,
using a timer works on real mode (simulated ERROR cases where the server is killed),
but not on Travis, and not because of GracefulStop but because the timer itself, Stop / GracefulStop doesnt finish when running in test mode from what i see.

we can add a variable to the function, that will determine if to use timer or not,
or to split this PR to 2 parts:
1. just change to GracefulStop, because it cause tests to be flaky
2. add the timer and fix Travis problem in a different method

Wdyt please ?

Reply:
added and removed, because the above problem, unless you suggest otherwise of course
/hold cancel

Reply:
I would prefer if we have the unit tests and the timeout.

Reply:
ack

Reply:
What you describe sounds to me like you just have to call `Stop()` after the test?

Reply:
It already calls it, i found that it just take a lot of time until Stop (focusing first on stop and then would move to Graceful stop),
detects all the `handleRawConn` are done.

Reply:
It seems now the unit tests are passing. Can you please explain what you did to make it pass?

Reply:
as discussed before, in my view from what i understood from it, the following approach is good to go for now:
vm_test create and destroy the server each loop, there is a closing of it, but the closing takes time, and is done in a goroutine which runs the server, and gets stop signal.
it takes time because it try to 
```
	// Finish handshaking (HTTP2)
	st := s.newHTTP2Transport(conn, authInfo)
	if st == nil {
		return
	}
```
which take very long, we might need a client to respone to it in order to make it short.

waiting enough at the test end shows it does stop them eventually, and we do asking them to be deleted,
but if we would wait for all to be closed, it would take more than 10 minutes which is limitation that the test framework enforce.

i would suggest that in another PR we would investigate even if we can remove the recreation of the severs, 
because it is the right thing to do, unless each test want to be totally standalone, even by its own "virt-handler".
and then just create the server one time or just create the socket might be enough.

Reply:
I'm sorry but I still don't get it. You wrote - "using a timer works on real mode (simulated ERROR cases where the server is killed),
but not on Travis, and not because of GracefulStop but because the timer itself, Stop / GracefulStop doesnt finish when running in test mode from what i see."

Now I see it works and travis passes. Did you change anything to make it work?

You also wrote "waiting enough at the test end shows it does stop *them* eventually, and we do asking *them* to be deleted,
but if we would wait for all to be closed, it would take more than 10 minutes which is limitation that the test framework enforce." - what are you talking about when you say "them"?

Reply:
> I'm sorry but I still don't get it. You wrote - "using a timer works on real mode (simulated ERROR cases where the server is killed),
> but not on Travis, and not because of GracefulStop but because the timer itself, Stop / GracefulStop doesnt finish when running in test mode from what i see."
> 
> Now I see it works and travis passes. Did you change anything to make it work?
> 

i removed the panic that was instead of Stop fallback in this PR, this is what i changed to make it work, since it actually takes time for the grpc server to stop, it takes time at `newHTTP2Transport`

> You also wrote "waiting enough at the test end shows it does stop _them_ eventually, and we do asking _them_ to be deleted,
> but if we would wait for all to be closed, it would take more than 10 minutes which is limitation that the test framework enforce." - what are you talking about when you say "them"?

them = the grpc servers that the test create delete in one of its contexts, per `It`

Reply:
thanks